### ISSUE TITLE
Tune whether to use mm or bmm for matmul in inductor max-autotune

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#118774]
Revert sgtm. As to how to approach this issue, on eager there is so much we can do, but this is something we can certainly tune in `torch.compile` with max-autotune.  cc @eellison @shunting314

_Originally posted by @lezcano in https://github.com/pytorch/pytorch/issues/118548#issuecomment-1916481836_

**Comment by xuzhao9** [comment]
I think bmm's poor performance when B is large is something we should look at or report to NVIDIA. All the above tests have the same computation FLOPS (BxMxNxK), but for aten::bmm the worst case (B=2048) is 14x slower than B=4.

**Comment by lezcano** [comment]
@xwang233 @IvanYashchuk Hi Xiao and Ivan! Is this something that the nvidia team would be interested in looking into?

**Comment by IvanYashchuk** [comment]
I think it's right to assume that batched matmul should have the same performance as folding batch dimensions and do a single large matmul. We shouldn't wait for the problem to be resolved in cuBLAS, we could implement our heuristic for `aten::bmm`.

cc: @eqy

**Comment by xuzhao9** [comment]
Is this also related to https://github.com/pytorch/pytorch/issues/114911 ?

**Comment by eqy** [comment]
@IvanYashchuk 

If I'm not misunderstanding, this would be something we could and should (maybe optionally) want to handle in eager-mode, correct? (Despite this surfacing from an apparent inductor use-case)

I can take the AI to cook something up if that's the case.

**Comment by eqy** [comment]
> Is this also related to #114911 ?

@xuzhao9 That is a separate cuBLAS issue due to broken workspace handling in cuBLAS, which should be fixed in an upcoming release.

**Comment by lezcano** [comment]
Yes, this is something we'd also want to do in eager mode, yep! (even if this issue is about `torch.compile`)

**Comment by xuzhao9** [comment]
@eqy  I have one question: the bmm kernel `cutlass::Kernel<cutlass_80_tensorop_f16_s16816gemm_relu_f16_256x64_32x4_nn_align8>` is a cutlass kernel, can we find its source code to perform deeper analysis?

We profiled it with NCU when input size is [4096, 4, 768, 768]:

BMM `cutlass::Kernel<cutlass_80_tensorop_f16_s16816gemm_relu_f16_256x64_32x4_nn_align8>`
<img width="1626" alt="image" src="https://github.com/pytorch/pytorch/assets/502017/da115254-49aa-4075-8189-e1879c1a6ad4">

There are a couple of strange things we noticed:
1. The device memory transfer is very high (4.86GB), much larger than the MM kernel `ampere_fp16_s16816gemm_fp16_128x128_ldg8_f2f_stages_32x5_nn`.
2. The L1 Cache hit rate is 0%, and L2 Cache hit rate is merely 25.69%. Looks like only L2 cache and shared memory are used.
3. The shared memory bank conflict rate is quite high.

**Comment by eqy** [comment]
@xuzhao9 Could you clarify what the dimensions correspond to B=4096? K=768? M=4?

The comment I got from @hwu36 is that L1 is bypassed due to limited reuse of the data, and that the bank conflicts metric combines dynamic + static bank conflicts and is therefore unhelpful.

**Comment by xuzhao9** [comment]
@eqy Yes, the dimensions are [4096, 4, 768] x [4096, 768, 768], therefore B=4096, M=4, N=K=768.

**Comment by hwu36** [comment]
if you M is just 4.  i don't understand why cublas let you run 256x64 tile size.  you only use 4/256 of the resource.

cutlass repository is https://github.com/NVIDIA/cutlass.  you can follow https://github.com/NVIDIA/cutlass/blob/main/media/docs/profiler.md to build cutlass profiler to profile this kernel.  the kernel name in cutlass is cutlass_tensorop_f16_s16816gemm_f16_256x64_32x4_nn_align8

cublas may transpose the problem, so you may also want to check cutlass_tensorop_f16_s16816gemm_f16_64x256_32x4_nn_align8

**Comment by xuzhao9** [comment]
@hwu36 If so, the problem might be why pytorch dispatches aten::bmm(4096, 4, 768, 768) to kernel `cutlass_80_tensorop_f16_s16816gemm_relu_f16_256x64_32x4_nn_align8`. I thought in eager mode, the kernel dispatch by input size is hardcoded?
